linera-core/src/client/mod.rs

```diff
@@ -1542,40 +1533,77 @@ where
         &self,
-        blob_id: BlobId,
+        blob_ids: Vec<BlobId>,
         remote_node: &RemoteNode<P::Node>,
     ) -> Result<(), ChainClientError> {
-        let certificate = remote_node.download_certificate_for_blob(blob_id).await?;
-        // This will download all ancestors of the certificate and process all of them locally.
-        self.receive_certificate(certificate).await?;
+        try_join_all(blob_ids.into_iter().map(|blob_id| async move {
+            let certificate = remote_node.download_certificate_for_blob(blob_id).await?;
+            // This will download all ancestors of the certificate and process all of them locally.
+            self.receive_certificate(certificate).await
+        }))
+        .await?;
+
         Ok(())
     }
 
     /// Downloads and processes a confirmed block certificate that uses the given blob.
     /// If this succeeds, the blob will be in our storage.
-    async fn receive_certificate_for_blob(&self, blob_id: BlobId) -> Result<(), ChainClientError> {
+    pub async fn receive_certificate_for_blob(
+        &self,
+        blob_id: BlobId,
+    ) -> Result<(), ChainClientError> {
+        self.receive_certificates_for_blobs(vec![blob_id]).await
+    }
+
+    /// Downloads and processes confirmed block certificates that use the given blobs.
+    /// If this succeeds, the blobs will be in our storage.
+    pub async fn receive_certificates_for_blobs(
+        &self,
+        blob_ids: Vec<BlobId>,
+    ) -> Result<(), ChainClientError> {
         let validators = self.validator_nodes().await?;
-        let mut tasks = FuturesUnordered::new();
-        for remote_node in validators {
-            let remote_node = remote_node.clone();
-            tasks.push(async move {
-                let cert = remote_node
-                    .download_certificate_for_blob(blob_id)
-                    .await
-                    .ok()?;
-                self.receive_certificate(cert).await.ok()
-            });
+        let mut tasks = BTreeMap::new();
+
+        for blob_id in blob_ids {
+            if tasks.contains_key(&blob_id) {
+                continue;
+            }
+
+            tasks.insert(blob_id, FuturesUnordered::new());
+            for remote_node in validators.clone() {
+                let remote_node = remote_node.clone();
+                tasks[&blob_id].push(async move {
+                    let cert = remote_node
+                        .download_certificate_for_blob(blob_id)
+                        .await
+                        .ok()?;
+                    self.receive_certificate(cert).await.ok()
+                });
+            }
         }
 
-        while let Some(result) = tasks.next().await {
-            if result.is_some() {
-                return Ok(());
+        let mut missing_blobs = Vec::new();
+        for (blob_id, mut blob_id_tasks) in tasks {
+            let mut found_blob = false;
+            while let Some(result) = blob_id_tasks.next().await {
```

Comment:
Should we shuffle the tasks here? Otherwise it's always the same validator we're waiting for first.